tests/loki_test.py

```diff
@@ -0,0 +1,197 @@
+        headers["X-Grafana-API-Key"] = key
+    return headers
+
+
+@pytest_asyncio.fixture(loop_scope="session", scope="session")
+async def mcp_client(mcp_url, grafana_headers):
+    async with sse_client(mcp_url, headers=grafana_headers) as (
+        read,
+        write,
+    ):
+        async with ClientSession(read, write) as session:
+            # Initialize the connection
+            await session.initialize()
+            yield session
+
+
+@pytest.mark.asyncio(loop_scope="session")
+@pytest.mark.parametrize("model", models)
+async def test_loki(model: str, mcp_client: ClientSession):
+    tools = await mcp_client.list_tools()
+    prompt = "what are the most recent log lines coming from Grafana?"
+
+    messages: list[Message] = [
+        Message(role="system", content="You are a helpful assistant."),  # type: ignore
+        Message(role="user", content=prompt),  # type: ignore
+    ]
+    tools = [convert_tool(t) for t in tools.tools]
+
+    response = await acompletion(
+        model=model,
+        messages=messages,
+        tools=tools,
+    )
+
+    # Check that there's a tool call.
+    assert isinstance(response, ModelResponse)
+    messages.extend(
+        await assert_and_handle_tool_call(response, mcp_client, "list_datasources")
+    )
+
+    # Call the LLM including the tool call result.
+    response = await acompletion(
+        model=model,
+        messages=messages,
+        tools=tools,
+    )
+
+    # Check that there's a tool call.
+    assert isinstance(response, ModelResponse)
+    messages.extend(
+        await assert_and_handle_tool_call(
+            response,
+            mcp_client,
+            "list_loki_label_names",
+            {"datasourceUid": "loki"},
+        )
+    )
+
+    # Call the LLM including the tool call result.
+    response = await acompletion(
+        model=model,
+        messages=messages,
+        tools=tools,
+    )
+
+    # Check that there's a tool call.
+    assert isinstance(response, ModelResponse)
+    messages.extend(
+        await assert_and_handle_tool_call(
+            response,
+            mcp_client,
+            "list_loki_label_values",
+            {"datasourceUid": "loki", "labelName": "container"},
+        )
+    )
+
+    # Call the LLM including the tool call result.
+    response = await acompletion(
+        model=model,
+        messages=messages,
+        tools=tools,
+    )
+
+    # Check that there's another tool call.
+    assert isinstance(response, ModelResponse)
+    messages.extend(
+        await assert_and_handle_tool_call(
+            response,
+            mcp_client,
+            "query_loki_logs",
+            {
+                "datasourceUid": "loki",
+                "logql": 'container="grafana"',
+            },
+        )
+    )
+
+    # Call the LLM including the tool call result.
+    response = await acompletion(
+        model=model,
+        messages=messages,
+        tools=tools,
+    )
```

Comment:
This shouldn't be in the final test really, I just wanted to see how well stuff worked. If the model gets there without needing to call these exact tools then that's fine.